### ISSUE TITLE
Constituency parser is unreliable

### ISSUE THREAD

**Issue by darabos** [allenai/allennlp#3246]
The problem I observe is that the constituency parser gives different parse trees for inputs that don't have grammatical differences. Here's an example:

For `{Monkeys,Cats,Tigers} have a tail and four legs.` we get the same beautiful parse tree:
![image](https://user-images.githubusercontent.com/1268018/64871757-32cf1e00-d646-11e9-8422-0878f7059260.png)
![image](https://user-images.githubusercontent.com/1268018/64871769-3b275900-d646-11e9-8d20-f65729a2c73e.png)
![image](https://user-images.githubusercontent.com/1268018/64871786-411d3a00-d646-11e9-964a-904dbeb327ff.png)

---

But if we replace `tail` with `head`, we get different trees for each animal:
![image](https://user-images.githubusercontent.com/1268018/64871880-6f027e80-d646-11e9-9bd5-86a8ca2b9d92.png)
![image](https://user-images.githubusercontent.com/1268018/64871895-775ab980-d646-11e9-87e8-40da9e10ee54.png)
![image](https://user-images.githubusercontent.com/1268018/64871909-7f1a5e00-d646-11e9-9a9f-2746f9e4a612.png)

In my use-case this is a problem. I'm writing a game where I'm parsing user input. If something works with monkeys, I'd really like it to work with cats and tigers too. If something works with the tail, I'd like it to work with the head too.

The parser in general is great and I'm very grateful for it. I just wonder how I should deal with this issue and was hoping you could share some advice.

I think I understand what causes it. This is a neural model trained on a corpus. It has no idea that there is anything in common between the words `cat` and `tiger`, other than what it has seen about their usage in the corpus. Because they are featured in quite different sentences, there will always be some difference in the network outputs when I change one animal to the other.

(Is that correct?)

I have some ill-formed ideas for dealing with this.

- I could train the model a bit more with data augmentation. I would take the original corpus, but randomly change words to other words that are surely grammatically equivalent. E.g. I'd change `cats` to `tigers`, `David` to `Jonathan`, `banana` to `tomato`, `purple` to `yellow`. I could use WordNet or some other dataset to pick safe replacements. This would get the behavior for different words in an interchangeable group to converge.

- I could retrain the model from scratch while blinding it to many of the actual words. E.g. instead of `Cats have a tail and four legs.` it would get `[Animal, plural] have a [body part] and [numeral] [body part, plural].`. This would guarantee that the words in the group get identical behavior.

  (Or I could even go for `[NNS] have a [NN] and [CD] [NNS].`?)

- I could keep the current model and just normalize the input. All animals get replaced with monkeys. All body parts get replaced with tails. All proper names get replaced with Isabella. Super hacky, but it would also ensure consistency. (And I wouldn't need to learn how to train the model.)

Do you think any of those could work? How do other people use this constituency parser? In what applications is this not a problem?

**Comment by brendan-ai2** [comment]
Broadly your understanding of why the parser behaves differently given these inputs is correct. This isn't an easy problem to solve robustly.

How you considered try Spacy's dependency parser? It could be a bit more consistent here.

**Comment by darabos** [comment]
> Broadly your understanding of why the parser behaves differently given these inputs is correct. This isn't an easy problem to solve robustly.

Thanks! What is the typical application for constituency parsers? I'm afraid maybe I'm just using it entirely wrong.

> How you considered try Spacy's dependency parser? It could be a bit more consistent here.

Yes. A few months ago I started with SpaCy's parser. If I remember correctly it didn't give very good trees for longer sentences. I first switched to Stanford CoreNLP's parser, but then discovered AllenNLP. AllenNLP had similar or better parse trees (for my sentences) and a more permissive license.

After posting my question I've found that the team who originally published the parser used in AllenNLP have published a more recent parser at https://github.com/nikitakit/self-attentive-parser. So I've switched again! Seems pretty solid so far.
